torch/ao/quantization/quantize_pt2e.py

```diff
@@ -78,6 +79,7 @@ def convert_pt2e(
 
     pm = PassManager([PortNodeMetaForQDQ()])
     model = pm(model).graph_module
+    constant_fold(model)
```

Comment:
why are we constant folding everything? And why use util from inductor?

It seems pretty weird that this change is inside inductor. we should decouple from that or refactor inductor stuff into utils that can be used outside.

Plus, blanket const folding everything, including stuff not related to quantization that maybe there, does not seem right.

Reply:
yeah this is a bit tricky I feel. e.g. we could have weight -> transpose -> q -> dq as well, do you fold "transpose -> q" or not?

it is weird that we just want to constant fold quantize op and not others, I guess it seems clean to me to do this later in the step as well (in lowering) if it can not be done cleanly in the quantization step

Reply:
May I know the motivation of moving the constant folding to the quantization phase? Do we have a use case that one would save a quantized model and load it later for deployment? Then, the folding of weights to low-precision form can save the model size?

BTW, inductor's constant folding assumes that model params are marked as "freezing" with "get_attr" calls in the graph. Do we have those taken care of here too?

Reply:
yeah people are checking the size of the model after quantization + lowering to executorch and found that the size does not change right now

> inductor's constant folding assumes that model params are marked as "freezing" with "get_attr" calls in the graph. Do we have those taken care of here too?

not following this, could you clarify a bit?

Reply:
See here: https://github.com/pytorch/pytorch/blob/b3272b2c0081485140e0762023617cfa64a39c2b/torch/_inductor/freezing.py#L81

We need to know what inputs are model parameters according to the tracing context and then replace them with "get_attr" calls. Later on, the constant folding logic will work on those "get_attr" calls. Not sure what happens with the quantization path.

Reply:
here we just constant propagate `get_attr` nodes, it's not related to freezing I think

Reply:
also what happens to metadata on the new get_attr nodes?

Reply:
> yeah this is a bit tricky I feel. e.g. we could have weight -> transpose -> q -> dq as well, do you fold "transpose -> q" or not?

Jerry, yes you const fold through this. My point is handling only things related to quantization and not other things. If there were other transforms that user introduces before coming to quantization, making assumption on const folding those is not right.

Reply:
OK, sounds good, so the criteria is fold any constant that has a quantize op in the path then? I'd like to reuse inductor constant folding because it's the same thing and I don't want to reimplement it. I can see how I can refactor that to fit this requirement

Reply:
> also what happens to metadata on the new get_attr nodes?

good question, I'll check this